modules/src/ics04_channel/handler/chan_open_init.rs

```diff
@@ -0,0 +1,184 @@
+//! Protocol logic specific to ICS3 messages of type `MsgChannelOpenInit`.
+
+use crate::handler::{HandlerOutput, HandlerResult};
+use crate::ics04_channel::channel::{ChannelEnd, State};
+use crate::ics04_channel::context::ChannelReader;
+use crate::ics04_channel::error::{Error, Kind};
+use crate::ics04_channel::handler::ChannelEvent::ChanOpenInit;
+use crate::ics04_channel::handler::ChannelResult;
+use crate::ics04_channel::msgs::chan_open_init::MsgChannelOpenInit;
+
+pub(crate) fn process(
+    ctx: &dyn ChannelReader,
+    msg: MsgChannelOpenInit,
+) -> HandlerResult<ChannelResult, Error> {
+    let mut output = HandlerOutput::builder();
+
+    let pc_id = (msg.port_id().clone(), msg.channel_id().clone());
+
+    if ctx.channel_end(&pc_id).is_some() {
+        return Err(Kind::ChannelExistsAlready(msg.channel_id().clone()).into());
+    }
+    //TODO what about the port capabilities  ?
+
+    if msg.channel().connection_hops().len() != 1 {
+        return Err(Kind::InvalidConnectionHopsLength.into());
+    }
+
+    // An IBC connection running on the local (host) chain should exist.
+    if ctx
+        .connection_state(&msg.channel().connection_hops()[0])
+        .is_none()
+    {
+        return Err(Kind::MissingConnection(msg.channel().connection_hops()[0].clone()).into());
+    }
+
+    //TODO Check Version non Empty but not necessary coherent
+    if msg.channel().version() == "" {
```

Comment:
This can be rewritten as:

```suggestion
    if msg.channel().version().is_empty() {
```

(As advised by [Clippy](https://github.com/rust-lang/rust-clippy#usage): https://rust-lang.github.io/rust-clippy/master/#comparison_to_empty)